bot/utils/scheduling.py

```diff
@@ -29,46 +31,70 @@ def __init__(self):
+        """Unschedule the task identified by `task_id`."""
+        log.trace(f"{self.cog_name}: cancelling task #{task_id}...")
+        task = self._scheduled_tasks.get(task_id)
 
-        if task is None:
-            log.warning(f"{self.cog_name}: Failed to unschedule {task_id} (no task found).")
+        if not task:
+            log.warning(f"{self.cog_name}: failed to unschedule {task_id} (no task found).")
             return
 
         task.cancel()
-        log.debug(f"{self.cog_name}: unscheduled task #{task_id}.")
-        del self.scheduled_tasks[task_id]
+        del self._scheduled_tasks[task_id]
+
+        log.debug(f"{self.cog_name}: unscheduled task #{task_id} {id(task)}.")
 
+    def _task_done_callback(self, task_id: str, done_task: asyncio.Task) -> None:
+        """
+        Delete the task and raise its exception if one exists.
 
-def create_task(loop: asyncio.AbstractEventLoop, coro_or_future: Union[Coroutine, asyncio.Future]) -> asyncio.Task:
-    """Creates an asyncio.Task object from a coroutine or future object."""
-    task: asyncio.Task = asyncio.ensure_future(coro_or_future, loop=loop)
+        If `done_task` and the task associated with `task_id` are different, then the latter
+        will not be deleted. In this case, a new task was likely rescheduled with the same ID.
+        """
+        log.trace(f"{self.cog_name}: performing done callback for task #{task_id} {id(done_task)}.")
 
-    # Silently ignore exceptions in a callback (handles the CancelledError nonsense)
-    task.add_done_callback(_silent_exception)
-    return task
+        scheduled_task = self._scheduled_tasks.get(task_id)
 
+        if scheduled_task and done_task is scheduled_task:
+            # A task for the ID exists and its the same as the done task.
+            # Since this is the done callback, the task is already done so no need to cancel it.
+            log.trace(f"{self.cog_name}: deleting task #{task_id} {id(done_task)}.")
+            del self._scheduled_tasks[task_id]
+        elif scheduled_task:
+            # A new task was likely rescheduled with the same ID.
+            log.debug(
+                f"{self.cog_name}: the scheduled task #{task_id} {id(scheduled_task)} "
+                f"and the done task {id(done_task)} differ."
+            )
+        elif not done_task.cancelled():
+            log.warning(
+                f"{self.cog_name}: task #{task_id} not found while handling task {id(done_task)}! "
+                f"A task somehow got unscheduled improperly (i.e. deleted but not cancelled)."
+            )
 
-def _silent_exception(future: asyncio.Future) -> None:
-    """Suppress future's exception."""
-    with contextlib.suppress(Exception):
-        future.exception()
+        with contextlib.suppress(asyncio.CancelledError):
+            exception = done_task.exception()
+            # Raise the exception if one exists.
+            if exception:
+                raise exception
```

Comment:
Does it make sense to raise it or should it be logged instead? Haven't tested this yet. I am guessing that if raised, it will just show up as 

```
asyncio | ERROR | Task exception was never retrieved
```

If logged here, more context could be provided.

Reply:
Tested it and here is what the exception looks like when raised:

```
2020-02-27 10:09:00,551 | asyncio | ERROR | Exception in callback Scheduler._task_done_callback(34)(<Task finishe...Error('aaah')>)
handle: <Handle Scheduler._task_done_callback(34)(<Task finishe...Error('aaah')>)>
Traceback (most recent call last):
  File "/opt/pyenv/versions/3.7.5/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/home/mark/repos/python/bot-pydis/bot/utils/scheduling.py", line 100, in _task_done_callback
    raise exception
  File "/home/mark/repos/python/bot-pydis/bot/cogs/moderation/scheduler.py", line 434, in _scheduled_task
    await asyncio.shield(self.deactivate_infraction(infraction))
  File "/home/mark/repos/python/bot-pydis/bot/cogs/moderation/scheduler.py", line 310, in deactivate_infraction
    raise ValueError("aaah")
ValueError: aaah
```

If logged:

```
2020-02-27 10:11:42,925 | bot.utils.scheduling | ERROR | Error in #35 140115635934496!
Traceback (most recent call last):
  File "/home/mark/repos/python/bot-pydis/bot/cogs/moderation/scheduler.py", line 434, in _scheduled_task
    await asyncio.shield(self.deactivate_infraction(infraction))
  File "/home/mark/repos/python/bot-pydis/bot/cogs/moderation/scheduler.py", line 310, in deactivate_infraction
    raise ValueError("aaah")
ValueError: aaah
```

Reply:
Logging it ourselves looks nicer. However, a user wouldn't know that an error occurred either way. How could they be notified (if applicable i.e. if it was ultimately invoked by a user)?

Reply:
I think that user-specific feedback would be rather difficult to accomplish, since there could be a long time between the user action and the failing of scheduled task. It also creates the issue of logistics: How are we going to know what to do from within the generic "done callback"?

Another option would be that, for more critical schedulers, we add an additional, scheduler-specific callback that handles it in some way. For the moderation scheduler, this could be a callback that has the Member that created the task added in using partial that, if a task finished with an error, sends an alert to mod-log accompanied with a ping to the original actor.

I'm mainly just thinking aloud here, though.